File: _archive/nodejs_backend/routes/pro.js
// routes/pro.js
const { Router } = require('express');
const createError = require('http-errors');
const { dbGet, dbRun } = require('../utils/db');

/**
 * Фáбрика маршрутов для работы с PRO-подпиской
 * @param {import('sqlite3').Database} db - экземпляр подключения к SQLite
 * @param {import('winston').Logger} logger - экземпляр логгера
 */
function proRouter(db) {
  const router = Router();

  // Grant PRO: POST /api/grantPro (для команды бота /grantpro)
  router.post('/grantPro', async (req, res, next) => {
    try {
      const { userId, days } = req.body;
      if (!userId || typeof days === 'undefined') {
        return res.status(400).json({ success: false, error: 'userId and days are required' });
      }
      const durationDays = parseInt(days, 10);
      if (isNaN(durationDays) || durationDays < 1) {
        return res.status(400).json({ success: false, error: 'days must be a positive number' });
      }
      
      const now = Date.now();
      // Получаем текущее pro_end
      const row = await dbGet(db, 'SELECT "pro_end" FROM users WHERE "userId" = ?', [userId]);
      
      let baseTime = now;
      if (row && row.pro_end) {
        const existing = Date.parse(row.pro_end);
        if (!isNaN(existing) && existing > now) baseTime = existing;
      }
      const newEnd = new Date(baseTime + durationDays * 24 * 60 * 60 * 1000).toISOString();
      
      await dbRun(db, 'UPDATE users SET is_pro = 1, "pro_end" = ? WHERE "userId" = ?', [newEnd, userId]);
      
      console.log(`PRO granted for user ${userId}, new end: ${newEnd}`);
      res.json({ success: true, is_pro: 1, pro_end: newEnd, end: newEnd });
    } catch (err) {
      console.error('❌ /api/grantPro error:', err);
      next(err);
    }
  });

  // Upgrade to PRO: POST /api/pro/upgrade
  router.post('/upgrade', async (req, res, next) => {
    try {
      const { userId, durationDays } = req.body;
      if (!userId || typeof durationDays === 'undefined') {
        throw createError(400, 'userId and durationDays are required');
      }
      const now = Date.now();
      // Получаем текущее pro_end
      const row = await dbGet(db, 'SELECT "pro_end" FROM users WHERE "userId" = ?', [userId]);
      
      let baseTime = now;
      if (row && row.pro_end) {
        const existing = Date.parse(row.pro_end);
        if (!isNaN(existing) && existing > now) baseTime = existing;
      }
      const newEnd = new Date(baseTime + durationDays * 24 * 60 * 60 * 1000).toISOString();
      
      await dbRun(db, 'UPDATE users SET is_pro = 1, "pro_end" = ? WHERE "userId" = ?', [newEnd, userId]);
      
      console.log(`PRO upgraded for user ${userId}, new end: ${newEnd}`);
      res.json({ success: true, is_pro: 1, pro_end: newEnd });
    } catch (err) {
      next(err);
    }
  });

  // Get PRO status: GET /api/pro/status?userId=...
  router.get('/status', async (req, res, next) => {
    try {
      const { userId } = req.query;
      if (!userId) throw createError(400, 'userId is required');
      const row = await dbGet(db, 'SELECT is_pro, "pro_end" FROM users WHERE "userId" = ?', [userId]);
      if (!row) throw createError(404, 'User not found');
      res.json({ success: true, is_pro: row.is_pro, pro_end: row.pro_end });
    } catch (err) {
      next(err);
    }
  });

  // Cancel PRO subscription: POST /api/pro/cancel
  router.post('/cancel', async (req, res, next) => {
    try {
      const { userId } = req.body;
      if (!userId) throw createError(400, 'userId is required');
      await dbRun(db, 'UPDATE users SET is_pro = 0 WHERE "userId" = ?', [userId]);
      console.log(`PRO cancelled for user ${userId}`);
      res.json({ success: true });
    } catch (err) {
      next(err);
    }
  });

  return router;
}

module.exports = proRouter;

File: python_backend/routes/pro.py
"""
routes/pro.py
Роуты для работы с PRO-подпиской
"""
from fastapi import APIRouter, Query, HTTPException, Body
from typing import Dict
from pydantic import BaseModel
from db_utils import db_get, db_run
from datetime import datetime, timedelta

router = APIRouter()


class GrantProRequest(BaseModel):
    userId: str
    days: int


class UpgradeProRequest(BaseModel):
    userId: str
    durationDays: int


class CancelProRequest(BaseModel):
    userId: str


@router.post("/grantPro")
async def grant_pro(data: GrantProRequest):
    """Выдать PRO-подписку пользователю (для команды бота)"""
    if not data.userId or data.days < 1:
        raise HTTPException(status_code=400, detail="userId and days (positive) are required")
    
    now = datetime.now()
    
    # Получаем текущее pro_end
    row = await db_get('SELECT "pro_end" FROM users WHERE "userId" = ?', [data.userId])
    
    base_time = now
    if row and row.get("pro_end"):
        try:
            existing = datetime.fromisoformat(row["pro_end"].replace("Z", "+00:00"))
            if existing > now:
                base_time = existing
        except (ValueError, TypeError):
            pass
    
    new_end = (base_time + timedelta(days=data.days)).isoformat()
    
    # Получаем текущее количество суперлайков
    # В БД колонка называется superLikesCount (camelCase)!
    user_row = await db_get('SELECT "superLikesCount" FROM users WHERE "userId" = ?', [data.userId])
    current_super_likes = user_row.get("superLikesCount", 0) if user_row else 0
    
    # Если суперлайков 0 или отсутствуют, выделяем 3 при выдаче PRO
    if current_super_likes == 0 or current_super_likes is None:
        await db_run('UPDATE users SET is_pro = 1, "pro_end" = ?, "superLikesCount" = 3 WHERE "userId" = ?', 
                     [new_end, data.userId])
        print(f"PRO granted for user {data.userId}, new end: {new_end}, superLikesCount set to 3")
    else:
        await db_run('UPDATE users SET is_pro = 1, "pro_end" = ? WHERE "userId" = ?', [new_end, data.userId])
        print(f"PRO granted for user {data.userId}, new end: {new_end}, superLikesCount kept: {current_super_likes}")
    
    return {"success": True, "is_pro": 1, "pro_end": new_end, "end": new_end}


@router.post("/upgrade")
async def upgrade_pro(data: UpgradeProRequest):
    """Обновить PRO-подписку"""
    if not data.userId or data.durationDays < 1:
        raise HTTPException(status_code=400, detail="userId and durationDays (positive) are required")
    
    now = datetime.now()
    
    row = await db_get('SELECT "pro_end" FROM users WHERE "userId" = ?', [data.userId])
    
    base_time = now
    if row and row.get("pro_end"):
        try:
            existing = datetime.fromisoformat(row["pro_end"].replace("Z", "+00:00"))
            if existing > now:
                base_time = existing
        except (ValueError, TypeError):
            pass
    
    new_end = (base_time + timedelta(days=data.durationDays)).isoformat()
    
    # Получаем текущее количество суперлайков
    # В БД колонка называется superLikesCount (camelCase)!
    user_row = await db_get('SELECT "superLikesCount" FROM users WHERE "userId" = ?', [data.userId])
    current_super_likes = user_row.get("superLikesCount", 0) if user_row else 0
    
    # Если суперлайков 0 или отсутствуют, выделяем 3 при обновлении PRO
    if current_super_likes == 0 or current_super_likes is None:
        await db_run('UPDATE users SET is_pro = 1, "pro_end" = ?, "superLikesCount" = 3 WHERE "userId" = ?', 
                     [new_end, data.userId])
        print(f"PRO upgraded for user {data.userId}, new end: {new_end}, superLikesCount set to 3")
    else:
        await db_run('UPDATE users SET is_pro = 1, "pro_end" = ? WHERE "userId" = ?', [new_end, data.userId])
        print(f"PRO upgraded for user {data.userId}, new end: {new_end}, superLikesCount kept: {current_super_likes}")
    
    return {"success": True, "is_pro": 1, "pro_end": new_end}


@router.get("/status")
async def get_pro_status(userId: str = Query(..., description="ID пользователя")):
    """Получить статус PRO-подписки"""
    if not userId:
        raise HTTPException(status_code=400, detail="userId is required")
    
    row = await db_get('SELECT is_pro, "pro_end" FROM users WHERE "userId" = ?', [userId])
    if not row:
        raise HTTPException(status_code=404, detail="User not found")
    
    return {"success": True, "is_pro": row.get("is_pro", 0), "pro_end": row.get("pro_end")}


@router.post("/cancel")
async def cancel_pro(data: CancelProRequest):
    """Отменить PRO-подписку"""
    if not data.userId:
        raise HTTPException(status_code=400, detail="userId is required")
    
    await db_run('UPDATE users SET is_pro = 0 WHERE "userId" = ?', [data.userId])
    print(f"PRO cancelled for user {data.userId}")
    return {"success": True}


class ActivatePromoCodeRequest(BaseModel):
    userId: str
    promoCode: str


@router.post("/activatePromoCode")
async def activate_promo_code(data: ActivatePromoCodeRequest):
    """Активировать промокод для пользователя"""
    if not data.userId or not data.promoCode:
        raise HTTPException(status_code=400, detail="userId and promoCode are required")
    
    promo_code = data.promoCode.strip().upper()
    
    # Проверяем существование промокода и его активность
    promo_row = await db_get(
        'SELECT id, days, is_active, expires_at FROM promo_codes WHERE code = ?',
        [promo_code]
    )
    
    if not promo_row:
        return {"success": False, "error": "Промокод не найден"}
    
    if not promo_row.get("is_active"):
        return {"success": False, "error": "Промокод неактивен"}
    
    # Проверяем срок действия промокода
    expires_at = promo_row.get("expires_at")
    if expires_at:
        from datetime import datetime, timezone
        now = datetime.now(timezone.utc)
        if isinstance(expires_at, str):
            expires_at = datetime.fromisoformat(expires_at.replace("Z", "+00:00"))
        if expires_at < now:
            return {"success": False, "error": "Промокод истек"}
    
    promo_code_id = promo_row["id"]
    days = promo_row["days"]
    
    # Проверяем, не использовал ли пользователь этот промокод ранее
    usage_row = await db_get(
        'SELECT id FROM promo_code_usage WHERE promo_code_id = ? AND user_id = ?',
        [promo_code_id, data.userId]
    )
    
    if usage_row:
        return {"success": False, "error": "Вы уже использовали этот промокод"}
    
    # Активируем PRO подписку
    now = datetime.now(timezone.utc)
    
    # Получаем текущее pro_end
    user_row = await db_get('SELECT "pro_end" FROM users WHERE "userId" = ?', [data.userId])
    
    base_time = now
    if user_row and user_row.get("pro_end"):
        try:
            existing = datetime.fromisoformat(user_row["pro_end"].replace("Z", "+00:00"))
            if existing > now:
                base_time = existing
        except (ValueError, TypeError):
            pass
    
    new_end = (base_time + timedelta(days=days)).isoformat()
    
    # Получаем текущее количество суперлайков
    user_row = await db_get('SELECT "superLikesCount" FROM users WHERE "userId" = ?', [data.userId])
    current_super_likes = user_row.get("superLikesCount", 0) if user_row else 0
    
    # Если суперлайков 0 или отсутствуют, выделяем 3 при активации PRO
    if current_super_likes == 0 or current_super_likes is None:
        await db_run(
            'UPDATE users SET is_pro = 1, "pro_end" = ?, "superLikesCount" = 3 WHERE "userId" = ?',
            [new_end, data.userId]
        )
    else:
        await db_run(
            'UPDATE users SET is_pro = 1, "pro_end" = ? WHERE "userId" = ?',
            [new_end, data.userId]
        )
    
    # Записываем использование промокода
    await db_run(
        'INSERT INTO promo_code_usage (promo_code_id, user_id) VALUES (?, ?)',
        [promo_code_id, data.userId]
    )
    
    print(f"✅ Промокод активирован: user_id={data.userId}, promo_code={promo_code}, days={days}, pro_end={new_end}")
    
    return {
        "success": True,
        "days": days,
        "pro_end": new_end,
        "message": f"✅ Промокод активирован! PRO подписка продлена на {days} дней."
    }

